lib/grpc/client/adapters/mint.ex

```diff
@@ -0,0 +1,182 @@
+defmodule GRPC.Client.Adapters.Mint do
+  @moduledoc """
+  A client adapter using mint
+  """
+
+  alias GRPC.{Channel, Credential}
+  alias GRPC.Client.Adapters.Mint.{ConnectionProcess, StreamResponseProcess}
+
+  @behaviour GRPC.Client.Adapter
+
+  @default_connect_opts [protocols: [:http2]]
+  @default_transport_opts [timeout: :infinity]
+
+  @impl true
+  def connect(%{host: host, port: port} = channel, opts \\ []) do
+    opts = Keyword.merge(@default_connect_opts, connect_opts(channel, opts))
+
+    channel
+    |> mint_scheme()
+    |> ConnectionProcess.start_link(host, port, opts)
+    |> case do
+      {:ok, pid} -> {:ok, %{channel | adapter_payload: %{conn_pid: pid}}}
+      # TODO add proper error handling
+      _error -> {:ok, %{channel | adapter_payload: %{conn_pid: nil}}}
+    end
+  end
+
+  @impl true
+  def disconnect(%{adapter_payload: %{conn_pid: pid}} = channel)
+      when is_pid(pid) do
+    ConnectionProcess.disconnect(pid)
+    {:ok, %{channel | adapter_payload: %{conn_pid: nil}}}
+  end
+
+  def disconnect(%{adapter_payload: %{conn_pid: nil}} = channel) do
+    {:ok, channel}
+  end
+
+  @impl true
+  def send_request(
+        %{channel: %{adapter_payload: %{conn_pid: pid}}, path: path} = stream,
+        message,
+        opts
+      )
+      when is_pid(pid) do
+    headers = GRPC.Transport.HTTP2.client_headers_without_reserved(stream, opts)
+    {:ok, data, _} = GRPC.Message.to_data(message, opts)
+
+    {:ok, stream_response_pid} =
+      StreamResponseProcess.start_link(stream, opts[:return_headers] || false)
+
+    response =
+      ConnectionProcess.request(pid, "POST", path, headers, data,
+        stream_response_pid: stream_response_pid
+      )
+
+    stream
+    |> GRPC.Client.Stream.put_payload(:response, response)
+    |> GRPC.Client.Stream.put_payload(:stream_response_pid, stream_response_pid)
+  end
+
+  @impl true
+  def receive_data(
+        %{server_stream: true, payload: %{response: response, stream_response_pid: pid}},
+        opts
+      ) do
+    with {:ok, headers} <- response do
+      stream = StreamResponseProcess.build_stream(pid)
+
+      case opts[:return_headers] do
+        true -> {:ok, stream, headers}
+        _any -> {:ok, stream}
+      end
+    end
+  end
+
+  # for streamed requests
+  def receive_data(
+        %{payload: %{response: {:ok, %{request_ref: _ref}}, stream_response_pid: pid}},
```

Comment:
why match on the ref with an empty variable?

Reply:
There are two usages for `receive_data/2`: 
- One is for server stream
- Other is for unary response

In this first one I'm patter matching for the server stream case. But it's a bit confusing in fact. I'll refactor this to return an atom, just to be clear what each match is solving. WDYT?

Reply:
I see! Perhaps you could use a `defguardp` to validate the `{:ok, value}`, like `defguardp is_client_stream_response(value) when is_ref(value.request_ref)` and another for `is_unary_request_response`

Reply:
WDYT of a guard for each case?

- `is_bidirectional_stream`
- `is_unary_request_stream_response`
- `is_stream_request_unary_response`
- `is_unary_request_response`

I think it would make it easier to understand responsibility of each case  + I could add a doc for each guard

Reply:
Actually I just refactor this, to use a cond. Way more cleaner than before.